tools/dragen-somatic/3.9.3/dragen-somatic__3.9.3.cwl

```diff
@@ -192,94 +310,70 @@ requirements:
           # Run dragen command and import options from cli
           $(get_dragen_eval_line())
 
-          # Check if --enable-map-align-output is set
-          if [[ ! "$(get_value_as_str(inputs.enable_map_align_output))" == "true" ]]; then
-            echo "--enable-map-align-output not set, no need to move normal bam file" 1>&2
-            echo "Exiting" 1>&2
-            exit
-          fi
+          # Check if fastq_list or fastq_list_rows is set
+          if [[ ! "$(get_value_as_str(inputs.fastq_list))" == "" ]] || [[! "$(get_value_as_str(inputs.fastq_list_rows))" == "" ]]; then
+            # Check if --enable-map-align-output is set
+            if [[ ! "$(get_value_as_str(inputs.enable_map_align_output))" == "true" ]]; then
+              echo "--enable-map-align-output not set, no need to move normal bam file" 1>&2
+              echo "Exiting" 1>&2
+              exit
+            fi
 
-          # Ensure that we have a normal RGSM value, otherwise exit.
-          if [[ -z "$(get_value_as_str(get_normal_name_from_fastq_list_csv()))" ]]; then
-            echo "Could not get the normal bam file prefix" 1>&2
-            echo "Exiting" 1>&2
-            exit
-          fi
+            # Ensure that we have a normal RGSM value, otherwise exit.
+            if [[ -z "$(get_value_as_str(get_normal_name_from_fastq_list_csv()))" ]]; then
```

Comment:
Looking at the function `get_normal_name_from_fastq_list_csv`, that will only work if the `inputs.fastq_list` is defined but doesn't cater for if `inputs.fastq_list_rows` is used instead.

Reply:
How does this look for `inputs.fastq_list_rows`:

```
- var get_normal_name_from_fastq_list_rows = function(){
          /*
          Get the normal sample name form  fastq list rows object
          */

          /*
          Check fastq list rows is defined
          */
          if (inputs.fastq_list_rows === null){
              return null;
          }

          /*
          Check contents are defined
          */
          if (inputs.fastq_list_rows.contents === null){
              return null;
          }

          /*
          Get RGSM value and return
          */
          rgsm_value = inputs.fastq_list_rows.map(function(value) {
              return value.rgsm[0];
          });

          /*
          rgsm is not in input. Return null
          */
          if (rgsm_value == null) {
              return null;
          }
        }
 ```

Reply:
Looking good, a couple of points:


## Part 1: 

I think we can take out the following (and it will always return null for the fastq list rows).
The contents attribute is for files which is important in the original `get_normal_name_from_fastq_list_csv` as the `fastq_list` attribute is a file object.  

```
/*
Check contents are defined
*/
if (inputs.fastq_list_rows.contents === null){
    return null;
}
```


## Part 2

If I emulate this part:

```
/*
Get RGSM value and return
*/
rgsm_value = inputs.fastq_list_rows.map(function(value) {
    return value.rgsm[0];
});
```

with

```
var inputs = new Object();

> inputs.fastq_list_rows = [
        { "rgsm": "foo" },
        { "rgsm": "foo" }
    ];

> rgsm_value = inputs.fastq_list_rows.map(function(value) {
        return value.rgsm[0];
    });
```

I get 
```
[ 'f', 'f' ]
```

`fastq_list_rows` is a list so this map with grab the 0th element of the `rgsm` attribute of each of the `fastq_list_rows` list. Rather than the `rgsm` attribute of the 0th element of `fastq_list_rows` list


Would this be simpler?
```
inputs.fastq_list_rows[0].rgsm
```

But I think once that's done we can easily add it into the original function with something like as so:

```
var get_normal_name_from_fastq_list_csv = function(){
    /*
    Try get the normal name from the fastq list rows:  
    */
    if (get_normal_name_from_fastq_list_rows_csv() !== null){
      return get_normal_name_from_fastq_list_rows_csv()
    }
    
    /*
    Get name from file instead
   */
    
    /*
    Check fastq list is defined
    */
    if (inputs.fastq_list === null){
        return null;
    }
    ...
}
```

Reply:
Gah, I added index to wrong variable. Probably copy pasted the wrong code chunk from the many things I found fascinating on internet. It should have been:

```
- var get_normal_name_from_fastq_list_rows = function(){
          /*
          Get the normal sample name form  fastq list rows object
          */

          /*
          Check fastq list rows is defined
          */
          if (inputs.fastq_list_rows === null){
              return null;
          }

          /*
          Get RGSM value and return
          */
          rgsm_value = inputs.fastq_list_rows.map(function(value) {
              return value.rgsm;
          });

          /*
          rgsm is not in input. Return null
          */
          if (rgsm_value == null) {
              return null;
          } else {
              return rgsm_value[0]
          }      
        }
```


> Would this be simpler? inputs.fastq_list_rows[0].rgsm

💯

Reply:
Updated in https://github.com/umccr/cwl-ica/commit/5c674919257220ce2ce3ffc6cd14afacc8963b7f